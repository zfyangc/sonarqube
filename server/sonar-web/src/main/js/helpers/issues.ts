/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { flatten, sortBy } from 'lodash';
import { SEVERITIES } from './constants';

interface TextRange {
  startLine: number;
  endLine: number;
  startOffset: number;
  endOffset: number;
}

interface FlowLocation {
  msg: string;
  textRange?: TextRange;
}

interface Comment {
  login: string;
  [x: string]: any;
}

interface User {
  login: string;
}

interface Rule {}

interface Component {}

interface IssueBase {
  severity: string;
  [x: string]: any;
}

export interface RawIssue extends IssueBase {
  assignee?: string;
  author?: string;
  comments?: Array<Comment>;
  component: string;
  flows?: Array<{
    locations?: FlowLocation[];
  }>;
  key: string;
  line?: number;
  project: string;
  rule: string;
  status: string;
  subProject?: string;
  textRange?: TextRange;
}

interface Issue extends IssueBase {}

export function sortBySeverity(issues: Issue[]): Issue[] {
  return sortBy(issues, issue => SEVERITIES.indexOf(issue.severity));
}

function injectRelational(
  issue: { [x: string]: any },
  source: any[] | undefined,
  baseField: string,
  lookupField: string
) {
  const newFields: { [x: string]: any } = {};
  const baseValue = issue[baseField];
  if (baseValue !== undefined && source !== undefined) {
    const lookupValue = source.find(candidate => candidate[lookupField] === baseValue);
    if (lookupValue != null) {
      Object.keys(lookupValue).forEach(key => {
        const newKey = baseField + key.charAt(0).toUpperCase() + key.slice(1);
        newFields[newKey] = lookupValue[key];
      });
    }
  }
  return newFields;
}

function injectCommentsRelational(issue: RawIssue, users?: User[]) {
  if (!issue.comments) {
    return {};
  }
  const comments = issue.comments.map(comment => {
    const commentWithAuthor = { ...comment, author: comment.login, login: undefined };
    return {
      ...commentWithAuthor,
      ...injectRelational(commentWithAuthor, users, 'author', 'login')
    };
  });
  return { comments };
}

function prepareClosed(issue: RawIssue) {
  return issue.status === 'CLOSED'
    ? { flows: undefined, line: undefined, textRange: undefined }
    : {};
}

function ensureTextRange(issue: RawIssue): { textRange?: TextRange } {
  return issue.line && !issue.textRange
    ? {
        textRange: {
          startLine: issue.line,
          endLine: issue.line,
          startOffset: 0,
          endOffset: 999999
        }
      }
    : {};
}

function reverseLocations(locations: FlowLocation[]): FlowLocation[] {
  const x = [...locations];
  x.reverse();
  return x;
}

function splitFlows(
  issue: RawIssue
): { secondaryLocations: FlowLocation[]; flows: FlowLocation[][] } {
  const parsedFlows = (issue.flows || [])
    .filter(flow => flow.locations != null)
    .map(flow => flow.locations!.filter(location => location.textRange != null));

  const onlySecondaryLocations = parsedFlows.every(flow => flow.length === 1);

  return onlySecondaryLocations
    ? { secondaryLocations: orderLocations(flatten(parsedFlows)), flows: [] }
    : { secondaryLocations: [], flows: parsedFlows.map(reverseLocations) };
}

function orderLocations(locations: FlowLocation[]) {
  return sortBy(
    locations,
    location => location.textRange && location.textRange.startLine,
    location => location.textRange && location.textRange.startOffset
  );
}

export function parseIssueFromResponse(
  issue: RawIssue,
  components?: Component[],
  users?: User[],
  rules?: Rule[]
): Issue {
  const { secondaryLocations, flows } = splitFlows(issue);
  return {
    ...issue,
    ...injectRelational(issue, components, 'component', 'key'),
    ...injectRelational(issue, components, 'project', 'key'),
    ...injectRelational(issue, components, 'subProject', 'key'),
    ...injectRelational(issue, rules, 'rule', 'key'),
    ...injectRelational(issue, users, 'assignee', 'login'),
    ...injectCommentsRelational(issue, users),
    ...prepareClosed(issue),
    ...ensureTextRange(issue),
    secondaryLocations,
    flows
  };
}
