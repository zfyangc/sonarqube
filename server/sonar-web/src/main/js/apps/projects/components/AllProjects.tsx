/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as PropTypes from 'prop-types';
import Helmet from 'react-helmet';
import PageHeader from './PageHeader';
import ProjectsList from './ProjectsList';
import PageSidebar from './PageSidebar';
import Visualizations from '../visualizations/Visualizations';
import { CurrentUser, isLoggedIn } from '../../../app/types';
import handleRequiredAuthentication from '../../../app/utils/handleRequiredAuthentication';
import ListFooter from '../../../components/controls/ListFooter';
import { translate } from '../../../helpers/l10n';
import * as storage from '../../../helpers/storage';
import { RawQuery } from '../../../helpers/query';
import '../styles.css';
import { Project, Facets } from '../types';
import { fetchProjects, parseSorting, SORTING_SWITCH } from '../utils';
import { parseUrlQuery, Query } from '../query';

export interface Props {
  currentUser: CurrentUser;
  isFavorite: boolean;
  location: { pathname: string; query: { [x: string]: string } };
  onSonarCloud: boolean;
  organization?: { key: string };
  organizationsEnabled: boolean;
}

interface State {
  facets?: Facets;
  loading: boolean;
  pageIndex?: number;
  projects?: Project[];
  query: Query;
  total?: number;
}

export default class AllProjects extends React.PureComponent<Props, State> {
  mounted: boolean;

  static contextTypes = {
    router: PropTypes.object.isRequired
  };

  constructor(props: Props) {
    super(props);
    this.state = { loading: true, query: {} };
  }

  componentDidMount() {
    this.mounted = true;

    const html = document.querySelector('html');
    if (html) {
      html.classList.add('dashboard-page');
    }

    if (this.props.isFavorite && !isLoggedIn(this.props.currentUser)) {
      handleRequiredAuthentication();
      return;
    }
    this.handleQueryChange(true);
    const footer = document.getElementById('footer');
    if (footer) {
      footer.classList.add('page-footer-with-sidebar');
    }
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps.location.query !== this.props.location.query) {
      this.handleQueryChange(false);
    }
  }

  componentWillUnmount() {
    this.mounted = false;

    const html = document.querySelector('html');
    if (html) {
      html.classList.remove('dashboard-page');
    }

    const footer = document.getElementById('footer');
    if (footer) {
      footer.classList.remove('page-footer-with-sidebar');
    }
  }

  getView = () => this.state.query.view || 'overall';

  getVisualization = () => this.state.query.visualization || 'risk';

  getSort = () => this.state.query.sort || 'name';

  isFiltered = (query = this.state.query) =>
    Object.values(query).some(value => value !== undefined);

  stopLoading = () => {
    if (this.mounted) {
      this.setState({ loading: false });
    }
  };

  fetchProjects = (query: any) => {
    this.setState({ loading: true, query });
    fetchProjects(
      query,
      this.props.isFavorite,
      this.props.organization && this.props.organization.key
    ).then(response => {
      if (this.mounted) {
        this.setState({
          facets: response.facets,
          loading: false,
          pageIndex: 1,
          projects: response.projects,
          total: response.total
        });
      }
    }, this.stopLoading);
  };

  fetchMoreProjects = () => {
    const { pageIndex, projects, query } = this.state;
    if (pageIndex && projects && query) {
      this.setState({ loading: true });
      fetchProjects(
        query,
        this.props.isFavorite,
        this.props.organization && this.props.organization.key,
        pageIndex + 1
      ).then(response => {
        if (this.mounted) {
          this.setState({
            loading: false,
            pageIndex: pageIndex + 1,
            projects: [...projects, ...response.projects]
          });
        }
      }, this.stopLoading);
    }
  };

  getSavedOptions = () => {
    const options: {
      sort?: string;
      view?: string;
      visualization?: string;
    } = {};
    if (storage.getSort()) {
      options.sort = storage.getSort() || undefined;
    }
    if (storage.getView()) {
      options.view = storage.getView() || undefined;
    }
    if (storage.getVisualization()) {
      options.visualization = storage.getVisualization() || undefined;
    }
    return options;
  };

  handlePerspectiveChange = ({ view, visualization }: { view: string; visualization?: string }) => {
    const query: {
      view: string | undefined;
      visualization: string | undefined;
      sort?: string | undefined;
    } = {
      view: view === 'overall' ? undefined : view,
      visualization
    };

    if (this.state.query.view === 'leak' || view === 'leak') {
      if (this.state.query.sort) {
        const sort = parseSorting(this.state.query.sort);
        if (SORTING_SWITCH[sort.sortValue]) {
          query.sort = (sort.sortDesc ? '-' : '') + SORTING_SWITCH[sort.sortValue];
        }
      }
      this.context.router.push({ pathname: this.props.location.pathname, query });
    } else {
      this.updateLocationQuery(query);
    }

    storage.saveSort(query.sort);
    storage.saveView(query.view);
    storage.saveVisualization(visualization);
  };

  handleSortChange = (sort: string, desc: boolean) => {
    const asString = (desc ? '-' : '') + sort;
    this.updateLocationQuery({ sort: asString });
    storage.saveSort(asString);
  };

  handleQueryChange(initialMount: boolean) {
    const query = parseUrlQuery(this.props.location.query);
    const savedOptions = this.getSavedOptions();
    const savedOptionsSet = savedOptions.sort || savedOptions.view || savedOptions.visualization;

    // if there is no filter, but there are saved preferences in the localStorage
    if (initialMount && !this.isFiltered(query) && savedOptionsSet) {
      this.context.router.replace({ pathname: this.props.location.pathname, query: savedOptions });
    } else {
      this.fetchProjects(query);
    }
  }

  updateLocationQuery = (newQuery: RawQuery) => {
    this.context.router.push({
      pathname: this.props.location.pathname,
      query: {
        ...this.props.location.query,
        ...newQuery
      }
    });
  };

  renderSide = () => (
    <div className="layout-page-side-outer">
      <div
        className="layout-page-side projects-page-side"
        style={{ top: this.props.organization ? 95 : 30 }}>
        <div className="layout-page-side-inner">
          <div className="layout-page-filters">
            <PageSidebar
              facets={this.state.facets}
              isFavorite={this.props.isFavorite}
              organization={this.props.organization}
              query={this.state.query}
              showFavoriteFilter={!this.props.onSonarCloud}
              view={this.getView()}
              visualization={this.getVisualization()}
            />
          </div>
        </div>
      </div>
    </div>
  );

  renderHeader = () => (
    <div className="layout-page-header-panel layout-page-main-header">
      <div className="layout-page-header-panel-inner layout-page-main-header-inner">
        <div className="layout-page-main-inner">
          <PageHeader
            currentUser={this.props.currentUser}
            isFavorite={this.props.isFavorite}
            loading={this.state.loading}
            onPerspectiveChange={this.handlePerspectiveChange}
            onSortChange={this.handleSortChange}
            organization={this.props.organization}
            projects={this.state.projects}
            query={this.state.query}
            selectedSort={this.getSort()}
            total={this.state.total}
            view={this.getView()}
            visualization={this.getVisualization()}
          />
        </div>
      </div>
    </div>
  );

  renderMain = () =>
    this.getView() === 'visualizations' ? (
      <div className="layout-page-main-inner">
        {this.state.projects && (
          <Visualizations
            displayOrganizations={!this.props.organization && this.props.organizationsEnabled}
            projects={this.state.projects}
            sort={this.state.query.sort}
            total={this.state.total}
            visualization={this.getVisualization()}
          />
        )}
      </div>
    ) : (
      <div className="layout-page-main-inner">
        {this.state.projects && (
          <ProjectsList
            cardType={this.getView()}
            isFavorite={this.props.isFavorite}
            isFiltered={this.isFiltered()}
            organization={this.props.organization}
            projects={this.state.projects}
            query={this.state.query}
          />
        )}
        <ListFooter
          count={this.state.projects !== undefined ? this.state.projects.length : 0}
          loadMore={this.fetchMoreProjects}
          ready={!this.state.loading}
          total={this.state.total !== undefined ? this.state.total : 0}
        />
      </div>
    );

  render() {
    return (
      <div className="layout-page projects-page" id="projects-page">
        <Helmet title={translate('projects.page')} />

        {this.renderSide()}

        <div className="layout-page-main projects-page-content">
          {this.renderHeader()}
          {this.renderMain()}
        </div>
      </div>
    );
  }
}
