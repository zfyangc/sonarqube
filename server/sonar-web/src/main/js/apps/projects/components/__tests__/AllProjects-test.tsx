/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* eslint-disable import/order */
import * as React from 'react';
import { mount, shallow } from 'enzyme';
import AllProjects, { Props } from '../AllProjects';
import { getView, saveSort, saveView, saveVisualization } from '../../../../helpers/storage';

jest.mock('../ProjectsList', () => ({
  // eslint-disable-next-line
  default: function ProjectsList() {
    return null;
  }
}));

jest.mock('../PageHeader', () => ({
  // eslint-disable-next-line
  default: function PageHeader() {
    return null;
  }
}));

jest.mock('../PageSidebar', () => ({
  // eslint-disable-next-line
  default: function PageSidebar() {
    return null;
  }
}));

jest.mock('../../utils', () => {
  const utils = require.requireActual('../../utils');
  utils.fetchProjects = jest.fn(() => Promise.resolve({ projects: [] }));
  return utils;
});

jest.mock('../../../../helpers/storage', () => ({
  getSort: () => null,
  getView: jest.fn(() => null),
  getVisualization: () => null,
  saveSort: jest.fn(),
  saveView: jest.fn(),
  saveVisualization: jest.fn()
}));

const fetchProjects = require('../../utils').fetchProjects as jest.Mock<any>;

beforeEach(() => {
  (getView as jest.Mock<any>).mockImplementation(() => null);
  (saveSort as jest.Mock<any>).mockClear();
  (saveView as jest.Mock<any>).mockClear();
  (saveVisualization as jest.Mock<any>).mockClear();
  fetchProjects.mockClear();
});

it('renders', () => {
  const wrapper = shallowRender();
  expect(wrapper).toMatchSnapshot();
  wrapper.setState({ query: { view: 'visualizations' } });
  expect(wrapper).toMatchSnapshot();
});

it('fetches projects', () => {
  mountRender();
  expect(fetchProjects).lastCalledWith(
    {
      coverage: undefined,
      duplications: undefined,
      gate: undefined,
      languages: undefined,
      maintainability: undefined,
      new_coverage: undefined,
      new_duplications: undefined,
      new_lines: undefined,
      new_maintainability: undefined,
      new_reliability: undefined,
      new_security: undefined,
      reliability: undefined,
      search: undefined,
      security: undefined,
      size: undefined,
      sort: undefined,
      tags: undefined,
      view: undefined,
      visualization: undefined
    },
    false,
    undefined
  );
});

it('redirects to the saved search', () => {
  (getView as jest.Mock<any>).mockImplementation(() => 'leak');
  const replace = jest.fn();
  mountRender({}, jest.fn(), replace);
  expect(replace).lastCalledWith({ pathname: '/projects', query: { view: 'leak' } });
});

it('changes sort', () => {
  const push = jest.fn();
  const wrapper = shallowRender({}, push);
  wrapper.find('PageHeader').prop<Function>('onSortChange')('size', false);
  expect(push).lastCalledWith({ pathname: '/projects', query: { sort: 'size' } });
  expect(saveSort).lastCalledWith('size');
});

it('changes perspective to leak', () => {
  const push = jest.fn();
  const wrapper = shallowRender({}, push);
  wrapper.find('PageHeader').prop<Function>('onPerspectiveChange')({ view: 'leak' });
  expect(push).lastCalledWith({
    pathname: '/projects',
    query: { view: 'leak', visualization: undefined }
  });
  expect(saveSort).lastCalledWith(undefined);
  expect(saveView).lastCalledWith('leak');
  expect(saveVisualization).lastCalledWith(undefined);
});

it('updates sorting when changing perspective from leak', () => {
  const push = jest.fn();
  const wrapper = shallowRender({}, push);
  wrapper.setState({ query: { sort: 'new_coverage', view: 'leak' } });
  wrapper.find('PageHeader').prop<Function>('onPerspectiveChange')({
    view: undefined
  });
  expect(push).lastCalledWith({
    pathname: '/projects',
    query: { sort: 'coverage', view: undefined, visualization: undefined }
  });
  expect(saveSort).lastCalledWith('coverage');
  expect(saveView).lastCalledWith(undefined);
  expect(saveVisualization).lastCalledWith(undefined);
});

it('changes perspective to risk visualization', () => {
  const push = jest.fn();
  const wrapper = shallowRender({}, push);
  wrapper.find('PageHeader').prop<Function>('onPerspectiveChange')({
    view: 'visualizations',
    visualization: 'risk'
  });
  expect(push).lastCalledWith({
    pathname: '/projects',
    query: { view: 'visualizations', visualization: 'risk' }
  });
  expect(saveSort).lastCalledWith(undefined);
  expect(saveView).lastCalledWith('visualizations');
  expect(saveVisualization).lastCalledWith('risk');
});

function mountRender(props: any = {}, push: Function = jest.fn(), replace: Function = jest.fn()) {
  return mount(
    <AllProjects
      currentUser={{ isLoggedIn: true }}
      fetchProjects={jest.fn()}
      isFavorite={false}
      location={{ pathname: '/projects', query: {} }}
      {...props}
    />,
    { context: { router: { push, replace } } }
  );
}

function shallowRender(
  props: Partial<Props> = {},
  push: Function = jest.fn(),
  replace: Function = jest.fn()
) {
  const wrapper = shallow(
    <AllProjects
      currentUser={{ isLoggedIn: true }}
      isFavorite={false}
      location={{ pathname: '/projects', query: {} }}
      onSonarCloud={false}
      organizationsEnabled={false}
      {...props}
    />,
    { context: { router: { push, replace } } }
  );
  wrapper.setState({
    loading: false,
    projects: [{ key: 'foo', measures: {}, name: 'Foo' }],
    total: 0
  });
  return wrapper;
}
