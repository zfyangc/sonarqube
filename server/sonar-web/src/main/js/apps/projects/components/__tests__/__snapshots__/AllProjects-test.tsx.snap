// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="layout-page projects-page"
  id="projects-page"
>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
    title="projects.page"
  />
  <ScreenPositionHelper
    className="layout-page-side-outer"
  />
  <div
    className="layout-page-main"
  >
    <div
      className="layout-page-header-panel layout-page-main-header"
    >
      <div
        className="layout-page-header-panel-inner layout-page-main-header-inner"
      >
        <div
          className="layout-page-main-inner"
        >
          <PageHeader
            currentUser={
              Object {
                "isLoggedIn": true,
              }
            }
            loading={false}
            onPerspectiveChange={[Function]}
            onQueryChange={[Function]}
            onSortChange={[Function]}
            projects={
              Array [
                Object {
                  "key": "foo",
                  "measures": Object {},
                  "name": "Foo",
                },
              ]
            }
            query={
              Object {
                "coverage": undefined,
                "duplications": undefined,
                "gate": undefined,
                "languages": undefined,
                "maintainability": undefined,
                "new_coverage": undefined,
                "new_duplications": undefined,
                "new_lines": undefined,
                "new_maintainability": undefined,
                "new_reliability": undefined,
                "new_security": undefined,
                "reliability": undefined,
                "search": undefined,
                "security": undefined,
                "size": undefined,
                "sort": undefined,
                "tags": undefined,
                "view": undefined,
                "visualization": undefined,
              }
            }
            selectedSort="name"
            total={0}
            view="overall"
            visualization="risk"
          />
        </div>
      </div>
    </div>
    <div
      className="layout-page-main-inner"
    >
      <ProjectsList
        cardType="overall"
        isFavorite={false}
        isFiltered={false}
        onSonarCloud={false}
        projects={
          Array [
            Object {
              "key": "foo",
              "measures": Object {},
              "name": "Foo",
            },
          ]
        }
        query={
          Object {
            "coverage": undefined,
            "duplications": undefined,
            "gate": undefined,
            "languages": undefined,
            "maintainability": undefined,
            "new_coverage": undefined,
            "new_duplications": undefined,
            "new_lines": undefined,
            "new_maintainability": undefined,
            "new_reliability": undefined,
            "new_security": undefined,
            "reliability": undefined,
            "search": undefined,
            "security": undefined,
            "size": undefined,
            "sort": undefined,
            "tags": undefined,
            "view": undefined,
            "visualization": undefined,
          }
        }
      />
      <ListFooter
        count={1}
        loadMore={[Function]}
        ready={true}
        total={0}
      />
    </div>
  </div>
</div>
`;

exports[`renders 2`] = `
<div
  className="layout-page projects-page"
  id="projects-page"
>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
    title="projects.page"
  />
  <ScreenPositionHelper
    className="layout-page-side-outer"
  />
  <div
    className="layout-page-main"
  >
    <div
      className="layout-page-header-panel layout-page-main-header"
    >
      <div
        className="layout-page-header-panel-inner layout-page-main-header-inner"
      >
        <div
          className="layout-page-main-inner"
        >
          <PageHeader
            currentUser={
              Object {
                "isLoggedIn": true,
              }
            }
            loading={false}
            onPerspectiveChange={[Function]}
            onQueryChange={[Function]}
            onSortChange={[Function]}
            projects={
              Array [
                Object {
                  "key": "foo",
                  "measures": Object {},
                  "name": "Foo",
                },
              ]
            }
            query={
              Object {
                "view": "visualizations",
              }
            }
            selectedSort="name"
            total={0}
            view="visualizations"
            visualization="risk"
          />
        </div>
      </div>
    </div>
    <div
      className="layout-page-main-inner"
    >
      <Visualizations
        displayOrganizations={false}
        projects={
          Array [
            Object {
              "key": "foo",
              "measures": Object {},
              "name": "Foo",
            },
          ]
        }
        total={0}
        visualization="risk"
      />
    </div>
  </div>
</div>
`;
